web/packages/teleterm/src/services/config/getKeyboardShortcutSchema.ts

```diff
@@ -0,0 +1,175 @@
+/**
+ * Copyright 2023 Gravitational, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { z } from 'zod';
+
+import { Platform } from 'teleterm/mainProcess/types';
+
+const VALID_SHORTCUT_MESSAGE =
+  'Valid shortcut can contain multiple modifiers (like "Shift") and a single key code (like "A" or "Tab"), combined by the "+" character.';
+
+export function invalidKeyCodeIssue(wrongKeyCode: string): z.IssueData {
+  return {
+    code: z.ZodIssueCode.custom,
+    message: `"${wrongKeyCode}" cannot be used as a key code. ${VALID_SHORTCUT_MESSAGE}`,
+  };
+}
+
+export function invalidModifierIssue(wrongModifier: string): z.IssueData {
+  return {
+    code: z.ZodIssueCode.custom,
+    message: `"${wrongModifier}" cannot be used as a modifier. ${VALID_SHORTCUT_MESSAGE}`,
+  };
+}
+
+export function duplicatedModifierIssue(): z.IssueData {
+  return {
+    code: z.ZodIssueCode.custom,
+    message: `Duplicated modifier found. ${VALID_SHORTCUT_MESSAGE}`,
+  };
+}
+
+export function getKeyboardShortcutSchema(platform: Platform) {
+  const allowedModifiers = getAllowedModifiers(platform);
+
+  return z
+    .string()
+    .transform(s => s.split('+'))
+    .transform(putModifiersFirst(allowedModifiers))
+    .superRefine(validateKeyCodeAndModifiers(allowedModifiers))
+    .transform(s => s.join('+'));
+}
+
+function putModifiersFirst(
+  allowedModifiers: string[]
+): (tokens: string[]) => string[] {
+  return tokens =>
+    tokens.sort((a, b) => {
+      if (allowedModifiers.indexOf(a) === -1) {
+        return 1;
+      }
+      if (allowedModifiers.indexOf(b) === -1) {
+        return -1;
+      }
+      return allowedModifiers.indexOf(a) - allowedModifiers.indexOf(b);
+    });
+}
+
+function validateKeyCodeAndModifiers(
+  allowedModifiers: string[]
+): (tokens: string[], ctx: z.RefinementCtx) => void {
+  return (tokens, ctx) => {
+    // empty accelerator disables the shortcut
+    if (tokens.join('') === '') {
+      return z.NEVER;
```

Comment:
What's the difference between returning `z.NEVER` and not calling `ctx.addIssue` (represented here) vs not returning `z.NEVER` and calling `ctx.addIssue` (represented in the last block)?

Reply:
> `NEVER` doesn't have any special meaning, I only followed the docs for aborting early https://github.com/colinhacks/zod#abort-early.

I took another look at the docs and they have this snippet:

```typescript
    // This is a special symbol you can use to
    // return early from the transform function.
    // It has type `never` so it does not affect the
    // inferred return type.
    return z.NEVER;
```